In [1]:
import os
import pandas as pd
from dotenv import load_dotenv
from sqlalchemy import create_engine, text

load_dotenv()

DB_HOST = os.getenv("DB_HOST")
DB_PORT = os.getenv("DB_PORT")
DB_USER = os.getenv("DB_USER")
DB_PASSWORD = os.getenv("DB_PASSWORD")
DB_NAME = os.getenv("DB_NAME")

engine = create_engine(
    f"mysql+pymysql://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}"
)

engine

Engine(mysql+pymysql://root:***@localhost:3307/classicmodels)

In [2]:
query = text("""
SELECT COLUMN_NAME, DATA_TYPE
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_SCHEMA = :db_name
  AND TABLE_NAME = 'customers'
""")

columns = pd.read_sql(
    query,
    engine,
    params={"db_name": DB_NAME}
)

columns

,COLUMN_NAME,DATA_TYPE
0,customerNumber,int
1,customerName,varchar
2,contactLastName,varchar
3,contactFirstName,varchar
4,phone,varchar
5,addressLine1,varchar
6,addressLine2,varchar
7,city,varchar
8,state,varchar
9,postalCode,varchar


In [3]:
def update_customer_phone(customer_number, new_phone):
    # Перевіряємо, чи існує клієнт
    check_query = text("""
        SELECT customerNumber, customerName, phone
        FROM customers
        WHERE customerNumber = :customer_number
    """)

    with engine.connect() as connection:
        customer = connection.execute(
            check_query,
            {"customer_number": customer_number}
        ).fetchone()

    if customer is None:
        print("Клієнта з таким номером не знайдено.")
        return

    # Оновлюємо телефон
    update_query = text("""
        UPDATE customers
        SET phone = :new_phone
        WHERE customerNumber = :customer_number
    """)

    with engine.begin() as connection:
        connection.execute(
            update_query,
            {
                "new_phone": new_phone,
                "customer_number": customer_number
            }
        )

    print("Телефон клієнта успішно оновлено.")

In [4]:
customers = pd.read_sql(
    text("""
        SELECT customerNumber, customerName, phone
        FROM customers
        LIMIT 10
    """),
    engine
)

customers

,customerNumber,customerName,phone
0,103,Atelier graphique,40.32.2555
1,112,Signal Gift Stores,7025551838
2,114,"Australian Collectors, Co.",03 9520 4555
3,119,La Rochelle Gifts,40.67.8555
4,121,Baane Mini Imports,07-98 9555
5,124,Mini Gifts Distributors Ltd.,4155551450
6,125,Havel & Zbyszek Co,(26) 642-7555
7,128,"Blauer See Auto, Co.",+49 69 66 90 2555
8,129,Mini Wheels Co.,6505555787
9,131,Land of Toys Inc.,2125557818


In [5]:
update_customer_phone(103, "050-123-4567")

Телефон клієнта успішно оновлено.


In [6]:
check_customer = pd.read_sql(
    text("""
        SELECT customerNumber, customerName, phone
        FROM customers
        WHERE customerNumber = :customer_number
    """),
    engine,
    params={"customer_number": 103}
)

check_customer


,customerNumber,customerName,phone
0,103,Atelier graphique,050-123-4567


In [7]:
products_test = pd.read_sql(
    text("""
        SELECT
            productCode,
            productName,
            quantityInStock,
            buyPrice
        FROM products
        ORDER BY productCode
        LIMIT 10
    """),
    engine
)

products_test

,productCode,productName,quantityInStock,buyPrice
0,S10_1678,1969 Harley Davidson Ultimate Chopper,7933,48.81
1,S10_1949,1952 Alpine Renault 1300,7305,98.58
2,S10_2016,1996 Moto Guzzi 1100i,6625,68.99
3,S10_4698,2003 Harley-Davidson Eagle Drag Bike,5582,91.02
4,S10_4757,1972 Alfa Romeo GTA,3252,85.68
5,S10_4962,1962 LanciaA Delta 16V,6791,103.42
6,S12_1099,1968 Ford Mustang,68,95.34
7,S12_1108,2001 Ferrari Enzo,3619,95.59
8,S12_1666,1958 Setra Bus,1579,77.90
9,S12_2823,2002 Suzuki XREO,9997,66.27


In [8]:
orders_columns = pd.read_sql(
    text("""
        SELECT COLUMN_NAME, DATA_TYPE
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA = 'classicmodels'
          AND TABLE_NAME = 'orders'
        ORDER BY ORDINAL_POSITION
    """),
    engine
)

orders_columns


,COLUMN_NAME,DATA_TYPE
0,orderNumber,int
1,orderDate,date
2,requiredDate,date
3,shippedDate,date
4,status,varchar
5,comments,text
6,customerNumber,int


In [9]:
orderdetails_columns = pd.read_sql(
    text("""
        SELECT COLUMN_NAME, DATA_TYPE
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA = 'classicmodels'
          AND TABLE_NAME = 'orderdetails'
        ORDER BY ORDINAL_POSITION
    """),
    engine
)

orderdetails_columns

,COLUMN_NAME,DATA_TYPE
0,orderNumber,int
1,productCode,varchar
2,quantityOrdered,int
3,priceEach,decimal
4,orderLineNumber,smallint


In [10]:
def create_order(customer_number, items):
    # Находим новый номер заказа
    with engine.connect() as connection:
        max_order = connection.execute(
            text("SELECT MAX(orderNumber) FROM orders")
        ).scalar()

    new_order_number = max_order + 1

    try:
        # Все действия внутри одной транзакции
        with engine.begin() as connection:

            # Проверяем, существует ли клиент
            customer = connection.execute(
                text("""
                    SELECT customerNumber
                    FROM customers
                    WHERE customerNumber = :customer_number
                """),
                {"customer_number": customer_number}
            ).fetchone()

            if customer is None:
                raise ValueError("Клієнта не знайдено.")

            # Проверяем наличие каждого товара на складе
            for item in items:
                product = connection.execute(
                    text("""
                        SELECT quantityInStock
                        FROM products
                        WHERE productCode = :product_code
                    """),
                    {"product_code": item["productCode"]}
                ).fetchone()

                if product is None:
                    raise ValueError(
                        f"Товар {item['productCode']} не знайдено."
                    )

                if product[0] < item["quantity"]:
                    raise ValueError(
                        f"Недостатньо товару {item['productCode']} на складі."
                    )

            # Создаем заказ
            connection.execute(
                text("""
                    INSERT INTO orders
                    (orderNumber, orderDate, requiredDate,
                     shippedDate, status, comments, customerNumber)
                    VALUES
                    (:order_number, CURDATE(),
                     DATE_ADD(CURDATE(), INTERVAL 7 DAY),
                     NULL, 'In Process', 'Test order',
                     :customer_number)
                """),
                {
                    "order_number": new_order_number,
                    "customer_number": customer_number
                }
            )

            # Добавляем товары и уменьшаем остатки
            for line_number, item in enumerate(items, start=1):

                connection.execute(
                    text("""
                        INSERT INTO orderdetails
                        (orderNumber, productCode, quantityOrdered,
                         priceEach, orderLineNumber)
                        VALUES
                        (:order_number, :product_code, :quantity,
                         :price, :line_number)
                    """),
                    {
                        "order_number": new_order_number,
                        "product_code": item["productCode"],
                        "quantity": item["quantity"],
                        "price": item["price"],
                        "line_number": line_number
                    }
                )

                connection.execute(
                    text("""
                        UPDATE products
                        SET quantityInStock = quantityInStock - :quantity
                        WHERE productCode = :product_code
                    """),
                    {
                        "quantity": item["quantity"],
                        "product_code": item["productCode"]
                    }
                )

        print(f"Замовлення №{new_order_number} успішно створено.")
        return new_order_number

    except Exception as e:
        print("Помилка. Транзакцію скасовано.")
        print(e)
        return None

In [11]:
test_items = [
    {
        "productCode": "S10_1678",
        "quantity": 2,
        "price": 48.81
    },
    {
        "productCode": "S10_1949",
        "quantity": 3,
        "price": 98.58
    }
]

new_order_number = create_order(103, test_items)


Замовлення №10426 успішно створено.


In [12]:
check_order = pd.read_sql(
    text("""
        SELECT
            o.orderNumber,
            o.orderDate,
            o.status,
            o.customerNumber,
            od.productCode,
            od.quantityOrdered,
            od.priceEach,
            od.orderLineNumber
        FROM orders AS o
        JOIN orderdetails AS od
            ON o.orderNumber = od.orderNumber
        WHERE o.orderNumber = :order_number
    """),
    engine,
    params={"order_number": new_order_number}
)

check_order

,orderNumber,orderDate,status,customerNumber,productCode,quantityOrdered,priceEach,orderLineNumber
0,10426,2026-09-16,In Process,103,S10_1678,2,48.81,1
1,10426,2026-09-16,In Process,103,S10_1949,3,98.58,2


In [13]:
check_stock = pd.read_sql(
    text("""
        SELECT
            productCode,
            productName,
            quantityInStock
        FROM products
        WHERE productCode IN ('S10_1678', 'S10_1949')
    """),
    engine
)

check_stock


,productCode,productName,quantityInStock
0,S10_1678,1969 Harley Davidson Ultimate Chopper,7931
1,S10_1949,1952 Alpine Renault 1300,7302


Висновок: Нове замовлення №10426 було успішно створено в межах однієї транзакції. Перед створенням замовлення було перевірено наявність товарів на складі. Товарні позиції додано до таблиці orderdetails, а кількість відповідних товарів у таблиці products зменшено на замовлену кількість. Результат перевірено за допомогою SELECT запитів.